### PR TITLE
add high-dpi module

### PR DESCRIPTION
**Pull request by Mic92**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Mic92** [comment]
Extracted from https://github.com/NixOS/nixos-hardware/pull/157

File: nixos/modules/hardware/video/high-dpi.nix [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    console.font = mkDefault
+      "${pkgs.terminus_font}/share/consolefonts/ter-u28n.psf.gz";
+
+    # Needed when typing in passwords for full disk encryption
+    console.earlySetup = mkDefault true;
+    boot.loader.systemd-boot.consoleMode = mkDefault "max";
+
+    services.xserver.dpi = mkDefault 196;
+    fonts.fontconfig.dpi = mkDefault 196;
+    environment.variables = mkDefault {
```
Comment (markuskowa) [inline nixos/modules/hardware/video/high-dpi.nix:18]: Would it make sense to also set a variable for QT: `QT_AUTO_SCREEN_SCALE_FACTOR=1` ?

Comment (Mic92) [reply nixos/modules/hardware/video/high-dpi.nix:18]: As far as I know QT now automatically detects the correct DPI.

**Review by markuskowa — APPROVED** [review]

**Comment by Mic92** [comment]
cc @gnidorah @colemickens can you give some feedback how this module would work on wayland? Right now I have only tested it on x11.

**Comment by gnidorah** [comment]
@Mic92 Doesn't work under sway, but sway has an own DPI settings. Perhaps @primeos could add more.

File: nixos/modules/hardware/video/high-dpi.nix [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  config = mkIf config.hardware.video.high-dpi.enable {
+    console.font = mkDefault
+      "${pkgs.terminus_font}/share/consolefonts/ter-u28n.psf.gz";
+
+    # Needed when typing in passwords for full disk encryption
+    console.earlySetup = mkDefault true;
+    boot.loader.systemd-boot.consoleMode = mkDefault "max";
+
+    services.xserver.dpi = mkDefault 196;
+    fonts.fontconfig.dpi = mkDefault 196;
```
Comment (Mic92) [inline nixos/modules/hardware/video/high-dpi.nix:17]: So I assume that everything starting from here should be only set, if x11 is enabled, or will wayland not be bothered by it?

Comment (gnidorah) [reply nixos/modules/hardware/video/high-dpi.nix:17]: @Mic92 Not bothered as I looked, and these should help XWayland programs

**Comment by emilazy** [comment]
I don't think overriding X11/fontconfig default DPI is a good idea. Lots of things act awkwardly or break with dpi ≠ 96 and, e.g. sway and GNOME kind of expect X11/fontconfig settings to be unscaled for Wayland HiDPI. Similarly, hardcoding `GDK_SCALE`/`_JAVA_OPTIONS` is inappropriate as many people will want fractional or 3× scaling and in general you shouldn't set `GDK_SCALE` if you don't have to (it breaks per-monitor scaling factors among other things).

The console font `latarcyrheb-sun32` is larger than `ter-u28n` (in fact it's the largest size the Linux console will support), and is already included in `kbd` (so avoids the `terminus_font` dependencies).

Is there a reason to set `boot.loader.systemd-boot.consoleMode`? UEFI firmware will generally initialize at the native resolution of the panel, and the default is to `keep` it.

File: nixos/modules/hardware/video/high-dpi.nix (hunk 2 above)
Comment (emilazy) [reply nixos/modules/hardware/video/high-dpi.nix:17]: Generally X11 is enabled even for pure-Wayland configurations (NixOS conflates "graphical environment" with "Xorg enabled" currently), so it's impossible to make configurations conditional on not using Wayland.

**Comment by matthewbauer** [comment]
I haven't found a good way to enable HiDPI *without* setting the `-dpi` flag. I don't understand why xserver can't detect my DPI from the EDID, I thought it was an Nvidia option, but perhaps it's intentional. It reports `Max Image Size [cm]: horiz.: 34  vert.: 19` on Dell XPS 9570 with Nvidia proprietary driver.

That said, I think we should have a clear warning in the description that this is a collection of hacks, and not really an ideal way to set DPI. For instance, if you connect an external monitor, I think it will use DPI=196 for it as well.

**Comment by emilazy** [comment]
That's because the X11 DPI setting is basically deprecated and the server just sets it to 96 no matter what, precisely because it wasn't a great solution for HiDPI. It can be useful as a hack for toolkits that don't support native HiDPI scaling, but those are increasingly rare these days, and it generally makes behaviour worse for the majority of applications using toolkits that do understand HiDPI.

If you're running without a desktop environment that manages the scaling settings for you, then your best bet is to set `QT_AUTO_SCREEN_SCALE_FACTOR`/`QT_SCREEN_SCALE_FACTORS`/`QT_SCALE_FACTOR`, `GDK_SCALE`, etc. and use various hacks for older toolkits (e.g. there are GTK+ 2 themes for HiDPI that just scale everything up by a fixed factor). The [ArchWiki page](https://wiki.archlinux.org/index.php/HiDPI) has a lot of good information, and it might be worthwhile to expose a twiddle to turn this stuff on to a fixed scaling factor for people who want simple single-output HiDPI support on X11 without a desktop environment, but it should probably not be the recommended HiDPI configuration or coupled to making console fonts readable at boot, since it'll make things worth for both the common cases and the cases that actually have the potential for best HiDPI support (GNOME, KDE, Wayland incl. sway, ...).

**Comment by matthewbauer** [comment]
> Edit: Sorry, missed your mention of Plasma... not finding any clear-cut information online, but I think it's worth twiddling with the QT_* variables to see if it gets things working for you.

So it looks like setting `QT_FONT_DPI=196` and `GDK_DPI_SCALE=1` are enough to get most apps working. Emacs still looks at xorg dpi though. I wish I didn't have to specify QT_FONT_DPI though, maybe this is a bug?

**Comment by Mic92** [comment]
> I don't think overriding X11/fontconfig default DPI is a good idea. Lots of things act awkwardly or break with dpi ≠ 96 and, e.g. sway and GNOME kind of expect X11/fontconfig settings to be unscaled for Wayland HiDPI. Similarly, hardcoding `GDK_SCALE`/`_JAVA_OPTIONS` is inappropriate as many people will want fractional or 3× scaling and in general you shouldn't set `GDK_SCALE` if you don't have to (it breaks per-monitor scaling factors among other things).

If they set a different scale by changing GDK_SCALE/_JAVA_OPTIONS it gets overriden, no? How to make it work without GDK_SCALE? I don't find it in the archlinux wiki. `gsettings` does not work outside of gnome afaik.

> 
> The console font `latarcyrheb-sun32` is larger than `ter-u28n` (in fact it's the largest size the Linux console will support), and is already included in `kbd` (so avoids the `terminus_font` dependency). It's upscaled, though; I forget if Terminus ships with 16x32 fonts but `spleen` does.

I used this font because it does not conflict with our [nixos-generate-config](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/installer/tools/nixos-generate-config.pl#L501).

**Comment by Mic92** [comment]
Maybe someone else can take this over. I just picked some settings, that worked for me.

**Comment by Mic92** [comment]
These settings are also used in the wiki: https://nixos.wiki/wiki/Xorg

**Comment by Mic92** [comment]
Here is a subset of the changes: https://github.com/NixOS/nixpkgs/pull/87730
The linux console stuff seems uncontroversial.

**Pull request closed without merging**